gh-118750: Asymptotically faster `int(string)`

tim-one opened the pull request:
Adding new, but unsued, `_dec_str_to_int_inner()`, + discussion.

<!-- gh-issue-number: gh-118750 -->
* Issue: gh-118750
<!-- /gh-issue-number -->

tim-one committed:
Adding new, but unsued, `_dec_str_to_int_inner()`, + discussion.

tim-one committed:
Correction: `decimal` in this conext computes reciprocals of powers

tim-one committed:
Clarify precedence in assert.

JelleZijlstra commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,55 @@ def inner(a, b):
+            if 0:
+                # This is maximally clear, but "too slow". `decimal`
+                # division is asymptotically fast, but we have no way to
+                # tell it to reuse the high-precision reciprocala it
```
```suggestion
                # tell it to reuse the high-precision reciprocal it
```

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

sstandre commented on Lib/_pylong.py:
```diff
@@ -361,3 +410,36 @@ def int_divmod(a, b):
+# compelling, though.
+#
+# If someone wants to look into speeding it more, I suggest focusing on
+# the multiplication by the cacned `5**(8**w2)`. We only need the
```
```suggestion
# the multiplication by the cached `5**(8**w2)`. We only need the
```

sstandre commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,55 @@ def inner(a, b):
+                # The only division in this alternative is by a power of
+                # 10, which comes nearly "for free" in decimal.
+                hi = n.scaleb(-8 * w2) # exactly n/10**(8*w2)
+                hi *= pow5to8[w2] # n/10**(8*w2) * 5**(8*w2( = n/2**(8*w2)
```
```suggestion
                hi *= pow5to8[w2] # n/10**(8*w2) * 5**(8*w2) = n/2**(8*w2)
```

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: sstandre <43125375+sstandre@users.noreply.github.com>

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: sstandre <43125375+sstandre@users.noreply.github.com>

tim-one committed:
Merge branch 'main' into str2int

tim-one committed:
Merge branch 'main' into str2int

tim-one committed:
My memory was wrong: using explicit, cached reciprocal approximations did very much better than I recalled. So moving to that instead. The crossover point is "only" about 3.4 million digits now, far smaller.

tim-one committed:
Merge branch 'str2int' of https://github.com/tim-one/cpython into str2int

tim-one committed:
Repair mysterious damage to comment.

tim-one committed:
Cut the precision of the reciprocal too before multiplying.

tim-one committed:
Merge branch 'main' into str2int

tim-one committed:
New comments.

tim-one committed:
Typo repair.

tim-one committed:
Raise BYTELIM.

Add a int<->str test for a truly large int (10 million digits),
which isn't currently tested. Bur regrtest will skip it unless
the "cpu" resource is enabled (e.g., via "-ucpu" on the cmdline).

tim-one committed:
For long strings, switch to the new implementation if `_decimal` is present.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

tim-one committed:
Update 2024-05-09-02-37-25.gh-issue-118750.7aLfT-.rst

Fix the news.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,91 @@ def inner(a, b):
+
+import math
+_LOG_10_BASE_256 = math.log(10, 256)
+del  math
```
`log10()` is more accurate.

Also, I think that you can simply use the precalculated constant. Even if on some platform a `float` can be calculated with more than 52-bit precision, this should not make much difference.

tim-one replied:
This isn't a base 10 logarithm - it's base 256. Don't want to plug in a literal string of digits. Subject to typing errors, copy/paste errors, and uncertainty about whether it's "right". Computing it instead is self-documenting, and at trivial cost (done once the first time the module is imported).

Don't really care about accuracy here. `w` will never get anywhere needing 53 bits of accuracy.

The far greater uncertainty is using the length of the string as a very crude guess about its base 10 logarithm. But don't care about that either :wink:.

serhiy-storchaka replied:
Then `math.log2(10) / 8` is more accurate.

tim-one replied:
On Windows, `math.log2(x) ` is generally _less_ accurate than `math.log(x, 2)`. Don't look at me - ask Microsoft :wink:.

Why do you care? It seems irrelevant to me. We're only retaining the integer part of the product, and unless the product is `> 2**53` all the fractional bits are thrown away, unlooked at. It doesn't matter if all of them are wrong.

serhiy-storchaka replied:
Isn't this a reason to use a more accurately precomputed value? Not that I care much about this.

tim-one replied:
I originally used `decimal` to compute the log to 20 decimal digits, and its `ln()` function is always correctly rounded to current precision. But then I thought "this is too silly to bear" :wink:. Newer code soon to be checked in punts if `w` requires at least 43 bits to express. So only a dead-broken `libm` could compute the log so badly as to matter.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,91 @@ def inner(a, b):
+        inner(hi, w - w2)
+        inner(lo, w2)
+
+    w = int(len(s) * _LOG_10_BASE_256) + 1
```
Did you try this with the significantly smaller or larger multiplier? For example 2 or 3? I think that the algorithm should still return the correct answer, but its time will be more than optimal. If `w` is too small, we will end with calling `int(str(n))` for much more than 512-digit values, which can even lead to recursion. If `w` is too large, there will be many `inner()` calls with `n = D(0)`. But please check yourself whether I am correct.

Testing this with such incorrect estimations for `w` can predict the behavior for very large `n`.

tim-one replied:
I tried making `w` itself 10x larger, and it only appeared to hurt speed. Cutting `w` in half, though, leads to the correction loop going around so often that its `assert` triggers (on the 10th iteration). I haven't yet dug into why that happened. Tomorrow :wink:.

tim-one replied:
So it's actually pretty obvious why. The working precision for the multiply is set high enough, but the precomputed reciprocal approximation hasn't retained enough precision to get the last digits of the integer part anywhere near being right (`n` is much larger than `pow256[w]`).

I think this is easily enough repaired, though, by forcing `w` to be an upper bound at the start.

tim-one replied:
Latest commit does force an upper bound. Thanks for pointing it out!

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,91 @@ def inner(a, b):
+            p256, recip = pow256[w2]
+            # The integer part will have about half the digits of n. So
+            # only need that much precision, plus some guard digits.
+            ctx.prec = (n.adjusted() >> 1) + 8
```
Did you try this with a smaller number of guard digits or without them at all? Does it still return the correct result (but slower)?

tim-one replied:
I have since replaced that code with a better method to pick the working precision, and with a much better explanation for "why" it's right.

Yes, I tried other values for guard digits. A lower bound on the quotient is always computed. The difference is in how many times the correction loop may need to go around. At 8, I've never seen it go around more than once, and even that for only 1 in about 100 thousand randomized inputs. It could be made smaller without harming speed, but it wouldn't measurably help speed either - 8 digits more or less is small in this context.

serhiy-storchaka replied:
I worry that for very large numbers it may require more than 1 correction, and that `assert count < 10` can fail. How smaller values for guard digits (1 or 2) affect the number of corrections? Is it always not more than 1 or 2? If it significantly grows, I would remove the `assert`.

tim-one replied:
I use `assert` for conditions I _believe_ are impossible to fail. But it's always the case that I may be wrong.

The number of digits shouldn't matter. This is about what can happen at the _last_ retained digit. With finite precision, it's always possible for the quotient to be too low by 1, but that can be made ever less likely by increasing the number of guard digits (so accumulated errors show up directly at _their_ low end instead).

Various analysis & coding errors while developing this convinced me that when the code is wrong, there's effectively no limit on how often the correction loop can go around. Better to raise an assert failure then than for the user's program to appear to hang using 100% CPU.

But, ya, with 8 guard digits I _believe_ 1 is the largest correction needed (no matter how large the input). I expect that with 7 guards too, or even 6. But below that, it gets fuzzier. As the newer code comments explain, it's the log10s of `n` and `p256` that matter to where the last integer-part digit is, and for both speed and simplicity I'm settling for `.adjusted()` as a very crude approximation to what those logs are. Guard digits are very effective at preventing bad consequences from incomplete analysis..

Indeed, I would have far less confidence in a purported "proof" of how badly the various approximations could end up hurting in the absence of guards.

tim-one replied:
Nope! Changed my mind again - 1 is the largest correction ever needed. But proof is involved. Short course: if `prec` is the decimal precision in effect, and we're rounding down, the result of an operation is exactly equal to the infinitely precise result times `1 - e` for some real `e` with `0 <= e < 10**(1-prec`). We have 3 operations: chopping `n` back to `prec` digits, likewise for `1/256**w2`, and also for their product.

So the computed product is exactly equal to the true product times `(1-e1)*(1-e2)*(1-e3)`; since the `e`s are all very small, an excellent approximation to the second factor is `1-(e1+e2+e3)` (the 2nd and 3rd order terms in the expanded product are too tiny to matter). If they're all as large as possible, that's `1 - 3*10**(1-prec)`. This, BTW, is all bog-standard FP error analysis.

Skipping details, that implies the computed product is within 1 of the true product provided `prec >= log10(true_product) + 1.47712125`.

`n.adjusted() - p256.adjusted()` is s crude integer approximation to `log10(true_product)` - but `prec` is necessarily an int too, so adding another guard digit or two is always enough to compensate for that. Again skipping details.

Also skipping why cutting the reciprocal to `p256.adjusted() + GUARD` digits to begin with is good enough. The precondition `n < 256**w` is needed to establish that the true product can't be too large for the reciprocal approximation to be too narrow.

And skipping other assorted details. I don't intend to change the code again anyway, or to write up such an involved proof in detail, but I'll probably add some high-level comments, akin to this one.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,91 @@ def inner(a, b):
+        inner(lo, w2)
+
+    w = int(len(s) * _LOG_10_BASE_256) + 1
+    with decimal.localcontext(_unbounded_dec_context) as ctx:
```
Does it create a copy of `_unbounded_dec_context` or returns a reference to `_unbounded_dec_context`?

tim-one replied:
It creates a copy, which the `with:` exit handler throws away. `_unbounded_dec_context` is not altered.

serhiy-storchaka replied:
Nice. So this code is not only thead-safe, but re-entrant.

serhiy-storchaka reviewed (commented):
This is a clever idea. It looks correct, I only have some questions about possible rounding errors. It may be that the code works for relatively small numbers (millions of digits), but fails for every millionth of qudrillion-digit numbers, for which we cannot realistically test. Testing the algorithm with intentionally worsen precision can allow to check its stability with lesser numbers.

tim-one committed:
Do a better job of picking working precision for the multiply.
And do a much better ;-) job of explaining it.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -219,7 +308,10 @@ def int_from_string(s):
     s = s.rstrip().replace('_', '')
-    return _str_to_int_inner(s)
+    func = _str_to_int_inner
+    if len(s) >= 3_500_000 and _decimal is not None:
```
On my computer I have equilibrium around 3_000_000. For 3_500_000 the new code is 20% faster, for 2_500_000 it is 20% slower.

So it is very close.

tim-one replied:
I see breakeven closer to 3.3.M, so 3.5M is a clear winner for both of us. "Do no harm" is more important to me than "aggressively pursue relatively small speedups" here, so I'll leave it alone.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Fotce the initial `w` to be an upper bound on the true value.

This is needed so the error analyais makes sense.

Also give up if `w` requires at least 43 bits, 10 less than
the precision of an IEEE-754 double. That's an extremely cautious
guard against low-order bit rot in log256(10) and *.

But I can't test that. Don't have 21 terabytes of RAM to build
a string long enough to trigger it. Nobody will ever, in good
faith, pass a string that large.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,117 @@ def inner(a, b):
+                # error analysis was fatally flawed in this case.
+                assert count < 10, (count, w, str(n))
+                lo -= p256
+                hi += 1
```
Can this loop be replaced with `divmod()`?
```py
            if lo >= p256:
                hi2, lo = divmod(lo, p256)
                hi += hi2
```
or few steps of subtraction followed by a `divmod()` if it was not enough?

tim-one replied:
It could be, but I don't want to. If the error analysis is wrong, we'll never find out about it then, but the code is nonetheless broken (perhaps not manifesting as wrong results, but in mysteriously inflated runtime at times).

This is the same, to me, as the earlier `assert lo >= 0`. _Analysis_ says it must be true - but I guard against mistaken analysis there too. All my "assert"s are of that nature. I do have less confidence in this one than in `lo >= 0`, though.

tim-one replied:
But I increasingly suspect I'm wrong. We're ignoring 3/4ths of the full "multiplication tree", and accounting for ever-more details of that now suggests the ignored part _can_ have significant effects on the integer part, and the larger the input the more "guard digits" may be needed (so that we build up more of the full multiplication tree) to move the possible errors back into the fractional part.

But finding crisp bounds on how bad it can get has eluded me, and changing the number of guard digits used depending on input size is unattractive (8 works fine - no more than 1 correction - for every input I tried so far, which have ranged up to 2.5 billion digits). So I expect to take your suggestion after all. Thanks!

BTW, I doubt more testing will find "a bad case". Even a "mere" 2.5 billion digit string takes well over an hour to convert. "Bad cases" would have a high percentage of large digits in both the input string and the reciprocal, ideally all 9s (but that's impossible for actual reciprocals). That makes ignoring the low-order digits in the full multiplication tree most damaging.

serhiy-storchaka reviewed (commented):
I do not understand your last change, but I think that ValueError is better than AssertionError. Or maybe OverflowError or RuntimeError are more suitable?

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Serhiy wore me out ;-) Add best possible log(10, 255) constant.

Since this eliminates the (by far) largest possible source of
error in computing the number of bytes needed, boost the
"string too large" limit to a `w` of 47 bits. Now we'd need a
339 terabyte input string to trigger it. Pure fantasy.
a

tim-one committed:
Repair comment.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Repair test for "close to maybe wrong" initial computation of `w`.

Change the correction-loop assert to put useful stuff in its
detail. Nobody was ever going to save such a message that
includes the full megabytes (or gigabytes, or ...) of the input
string. Instead give short, critical info, about the byte and
digit offets in play. That should suffice to reverse-engineer
what went wrong in the error analysis. _All_ input strings with
the same length go through exactly the same analysis.

Although, no, I still don't expect it to ever trigger.

erlend-aasland commented on Lib/_pylong.py:
```diff
@@ -210,6 +210,121 @@ def inner(a, b):
+    def inner(n, w):
+        #assert n < D256 ** w # required, but too expensive to check
+        if w <= BYTELIM:
+            # XXX Stefan Pochmann discovered that, for 1024-bit ints,
```
What does XXX imply here? Based on the successive sentence, I judge it is a kind of a TODO comment, but it is hard to tell. Mind spelling it out instead of adding a cryptic "XXX"?

tim-one replied:
It's been used in CPython's source for decades.

https://stackoverflow.com/questions/1452934/what-is-the-meaning-of-xxx-in-code-comments

A quick search on the `main` branch now found 976 instances in our `.py` files, 574 in `.c` files, and 131 in `.h` files. Far more common than "TODO". When in Rome .,,

Note that a PEP to impose rules on when to use which flavors of tag was rejected:

https://peps.python.org/pep-0350/

I'm going to close this too, because you're going to trip over a thousand other instances too - get used to it :smile:.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Added exact quotient correction for cases adding 1 once isn't enough.

We have no test case that takes that path, although they're easy to
provoke by reducing the new GUARD parameter & trying input strings
of the form `"9" * n`.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Check in testing helper I kept ripping out before earlier commits.
It's cheap, so leave it in.

And this allows adding a test using it to exercise the new
"failsafe" logic that guards against having too few guard digits.

tim-one committed:
And one more test, of thta absurdly large inputs are rejected.

tim-one committed:
Trying to repair new test that failed on some test platforms.

tim-one committed:
I don't know what the "WASI" testbot is, except that it's annoying ;-)

tim-one committed:
And more random thrashing.

tim-one committed:
Another random stab at making WASI happy.

tim-one committed:
More random thrashing to try to shut up WASI :-(

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Make GUARD a keyword-only argument. Plus many comment changes.

Explain GUARD in much more detail at the end of the file. I
believe 5 is actually large enough, but there's ne simple way
to explain why even a million would always be large enough.

tim-one committed:
repair spelling mistake in comment

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -361,3 +495,80 @@ def int_divmod(a, b):
+#
+# The default 8 is more than enough so no more than 1 correction step
+# was ever needed for all inputs tried through 2.5 billion digita. In
+# fact, I believe 5 guard digis are always enougn - but the proof is
```
```suggestion
# fact, I believe 5 guard digits are always enough - but the proof is
```

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -361,3 +495,80 @@ def int_divmod(a, b):
+# by a cached reciprocal approximation instead, fixing up errors
+# afterwards. This reduced the crossover point significantly,
+#
+# I revisited tha code, and found ways to improve and simplify it. The
```
```suggestion
# I revisited the code, and found ways to improve and simplify it. The
```

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,137 @@ def inner(a, b):
+#    print(float(mp.log(10, 256)).hex())
+_LOG_10_BASE_256 = float.fromhex('0x1.a934f0979a371p-2') # about 0.415
+
+# _apread is for internal testing. It maps a key to the number of times
```
```suggestion
# _spread is for internal testing. It maps a key to the number of times
```

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,137 @@ def inner(a, b):
+
+    # How many base 256 digits are needed?. Mathematically, exactly
+    # floor(log256(int(s))) + 1. There is no cheap way to compute this.
+    # But we can get an upper buond, and that's necessary for our error
```
```suggestion
    # But we can get an upper bound, and that's necessary for our error
```

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,137 @@ def inner(a, b):
+    log_ub = lenS * _LOG_10_BASE_256
+    log_ub_as_int = int(log_ub)
+    w = log_ub_as_int + 1 + (log_ub - log_ub_as_int > 0.9)
+    # And what if we'vv plain exhausted the limits of HW floats? We
```
```suggestion
    # And what if we have plain exhausted the limits of HW floats? We
```
?

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,137 @@ def inner(a, b):
+        # We're going to do inexact, chopped arithmetic, multiplying by
+        # an approximation to the reciprocal of 256**i. We chop to get a
+        # lower bound on the true integer quotient. Our approximation is
+        # a lower bound, the multiply is chopped too, and
```
```suggestion
        # a lower bound, the multiplier is chopped too, and
```
?

tim-one replied:
"multiply" is correct - it's the single most time-critical line in the code `hi = +n * +recip`. It's "chopped" because context precision was cut back on the preceding line. Otherwise the multiply would retain twice as many digits.

eendebakpt replied:
I see, multiplier is indeed not correct. In this case "multiplication" would read a bit more natural to me, maybe because I an not familiar with using the term "multiply" in this way.

tim-one replied:
Sure, happy to change it.

BTW, when you respond to  a "resolved conversation", please click the "unresolve conversation" button, else it's very easy to miss.

eendebakpt commented on Lib/test/test_int.py:
```diff
@@ -919,5 +919,57 @@ def test_pylong_roundtrip(self):
+
+    @unittest.skipUnless(_pylong, "pylong module required")
+    def test_whitebox_dec_str_to_int_inner_monster(self):
+        # I don't think anyone has enough RAM o build a string long enough
```
```suggestion
        # I don't think anyone has enough RAM to build a string long enough
```

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
I now believe GUARD=3(!) is sufficient. And that's the smallest
value the error analysis concluded _could_ work. The analysis
may actually be "crisp" now :-)

The problem wasn't ever in the "crude approximation" `.adjusted()`
gives for log10. Turns out that's always an upper bound on what's
needed.

The problem was that the precommputed reciprocal appromixations
didn't save quite enough digits to handle the worst cases. I got
burned by that before (and the fix _then_ was to force the guess
for the number of bytes needed to be an upper bound). This one
was subtler.

tim-one committed:
Apply suggestions from code review

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

tim-one committed:
The lint checker whinef about a blank line at the end of the file.
And, for whatever reason, GitHub appeared to go on to toss code
review changes committed after that but before this.

tim-one committed:
And restoring another code review change.

tim-one committed:
Update Lib/test/test_int.py

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Made suggested word change.

tim-one committed:
Just commehts: be more careful about explaining how bad the
approximation to log10(true_product) can be. It's actually
suprising how good it is, given how bad it is ;-)

tim-one committed:
Typo repair.

tim-one committed:
Someone should contribute a spell-checker to IDLE ;-)

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -277,44 +295,46 @@ def inner(n, w):
     # a fraction > 0.9.
-    log_ub = len(s) * _LOG_10_BASE_256
+
+    # The "WASI" test platfrom can complain about `len(s)` if it's too
+    # large to fit in its idea of "an index-sized integer".
+    lenS = s.__len__()
```
I think that it is a fair game to allow this to fail on WASI. In any case a ValueError is raised for too large input on other platforms. It will just be different (but perhaps close) limit on WASI, and possibly with different error type.

tim-one replied:
I'm strongly inclined not to change this again. It passes now, and the workarounds were easy in the end. Indeed, I'd rather remove the test entirely than waste more time on "WASI" :wink:.

serhiy-storchaka commented on Lib/_pylong.py:
```diff
@@ -277,44 +295,46 @@ def inner(n, w):
     log_ub_as_int = int(log_ub)
-    w = log_ub_as_int + 1 + (log_ub - log_ub_as_int > 9.9)
-    # And what if we'vv plain exhausted the limits of HW floats? We
+    w = log_ub_as_int + 1 + (log_ub - log_ub_as_int > 0.9)
```
Yes, the 9.9 constant confused me the most.

tim-one replied:
Sorry about that! The typo gave a test that could never succeed. The difference is always in the semi-open interval `[0.0, 1.0)`

serhiy-storchaka reviewed (approved):
Thanks for the clarifications, although I didn't fully understand them, as it is beyond my competence. When I asked you a question, it was primarily for you to clarify the matter for yourself, rather than for me. I'm glad if it helped.

And if you're happy with this code, I don't have any more questions that I can understand. Please do not forget to edit the commit message.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Free the memory for `hi` as soon as we're done with it.

This helps my testing on multi-gigabyte inputs, by
reducing peak RAM use.

tim-one committed:
Add notes about the details of how reciprocals are computed.

tim-one committed:
Repair grammar.

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -361,3 +496,124 @@ def int_divmod(a, b):
+# ------------------------
+# The code computes all the powers of 256 needed, and all those of
+# 1/256. These are exact, but the reciprocals have over twice as many
+# significant digts as needed. So in another pass we cut the exact
```
```suggestion
# significant digits as needed. So in another pass we cut the exact
```

eendebakpt commented on Lib/_pylong.py:
```diff
@@ -361,3 +496,124 @@ def int_divmod(a, b):
+# LATER: doing this analysis pointed out an error: our division isn't
+# exactly "balanced", in that when `w` is odd the integer part of
+# n/256**w2 can be larger than 256**w2. The code used enough working
+# precision in the multiply then, but the precommputed reciprocal
```
```suggestion
# precision in the multiply then, but the precomputed reciprocal
```

tim-one replied:
All methods inherit the asymptotics of bigint multiplication. The current method is throttled by CPython's Karatsuba multiplication. The exponent is `log2(3)`, due to that Karatsuba cleverly expresses a product as a combination of 3 half-width products. Double the input sizes, and schoolbook `*` takes 4 times longer; Karatsuba only 3x.

`decimal` implements Karatsuba too, but has a much fancier scheme beyond that. based on the Number Theoretic Transform (Google for NTT).  It's related to Fourier transforms. I don't know any details about this specific implementation, so can't give its asymptotics. Implementations are all over the map. "Naive NTT" is quadratic time, but the most sophisticated versions claim to be `O(n log n)`. From timing and some knowledge of its author, I'm sure this implementation is quite good indeed.

eendebakpt reviewed (commented):
Just two minor typos.

Out of curiosity: the old `_str_to_int_inner` has complexity O(len(s)**1.58) according to the docstring. Is there anything known about the complexity of the new method?

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Apply suggestions from code review

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

tim-one committed:
Merge branch 'str2int' of https://github.com/tim-one/cpython into str2int

tim-one committed:
Big change: keeping the exact reciprocals now.

No longer trying to truncate them "early". That was another
possible source of error that the error analysis didn't properly
account for, and one that bit me multiple times.

So screw it :-) It doesn't affect the speed, since the
recip is chopped back to `prec` digits before use anyway. The
difference now is that chopping will always retain as mamy
"good" digits as needed.

tim-one committed:
Remove no-longer-useful reset of ctx.prec.

tim-one committed:
Reciprocal approximations are back - you knew it was coming ;-)

Found a rigorous upper bound on the number of extra digits needed (4),
and reworked the error analysis to account for this too. I have no
_empirical_ evidence that 4 is needed, though, or even 3.

tim-one committed:
Comment repair.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Finished the proof - no more hand-waving on any point :-)

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Fix tiny typos in comments.

tim-one committed:
Fudge. Typo repair is adding new typos too :-(

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Consolidate all the seemingly random observations aboud .adjusted(),
in a self-contained new section that spells out everything needed.

tim-one committed:
typo

tim-one committed:
And we don't ever need exact reicrocals after all!
Unsure about why my protracted brain freeze about this.

tim-one committed:
Remove a mention of exact reciprocals, since they're no longer used.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Update comments. No code changes.

tim-one committed:
And exact reciprocals are back.
Can't beat them for speed, and that matters most at smaller
inputs. So it's either bring them back, or raise the crossover
point.

sstandre commented on Lib/_pylong.py:
```diff
@@ -211,6 +211,138 @@ def inner(a, b):
+            # 1/2**i has the same number of significant decimal digits
+            # as 5**i, generally over twice the number in 2**i,
+            ctx.prec = v.adjusted() + GUARD + 4
+            # The unary "+" chope the reciprocal back to that precision.
```
```suggestion
            # The unary "+" chops the reciprocal back to that precision.
```

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Split on ceiling(w/2) instead of on the floor.

This makes reasoning about reciprocal precision needed much
easier, and now nothing in the proofs depends on the output
base (256).

Alas, `compute_powere()` may, if its new `need_hi=True` argument
is specified, need to invoke `**` more than once now.

tim-one committed:
Delete a line of unused code.

tim-one committed:
Update Lib/_pylong.py

Co-authored-by: sstandre <43125375+sstandre@users.noreply.github.com>

tim-one committed:
Merge branch 'str2int' of https://github.com/tim-one/cpython into str2int

tim-one committed:
typo

tim-one committed:
Bah. Another typo.

tim-one committed:
Merge branch 'main' into str2int

tim-one committed:
Repair thinko in comment.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Splitting on ceiling(w/2) gave compute_powers() a harder job.
So make it smarter.

tim-one committed:
Merge branch 'main' into str2int

tim-one committed:
Reduce excessive indentation.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Give compute_powers() another IQ boost.

Also run millions of tests, but, at least for now, they need
to be enabled "by hand".

tim-one committed:
Add basic compoute_powers() test.

tim-one committed:
At least whan I make a typo, auto-commplete reproduces it ;-)

tim-one committed:
Fix old typo in comment everyone missed ;-)

tim-one committed:
Correct technical detail in comment.

tim-one committed:
And another stray typo :-(

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

nineteendo commented on Lib/_pylong.py:
```diff
@@ -211,6 +247,139 @@ def inner(a, b):
+    # And what if we've plain exhausted the limits of HW floats? We
+    # could compute the log to any desired precision using `decimal`,
+    # but it's not plausible that anyone will pass a string requiring
+    # trillions of bytes (unles they're just trying to "break things").
```
```suggestion
    # trillions of bytes (unless they're just trying to "break things").
```

tim-one replied:
Thank you for the careful eye!

nineteendo commented on Lib/_pylong.py:
```diff
@@ -211,6 +247,139 @@ def inner(a, b):
+        for k, v in pow256.items():
+            # No need to save much more precision in the reciprocal than
+            # the power of 256 has, plus some guard digits to absorb
+            # most relevant rounding errors. This is highly signficant:
```
```suggestion
            # most relevant rounding errors. This is highly significant:
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#
+# Lemma 2: ceiling(log10(x/y)) <= x.a - y.a + 1
+#
+# Peood: Express x and y as in Lemma 1. Then x/y = f/g * 10**(x.a -
```
```suggestion
# Proof: Express x and y as in Lemma 1. Then x/y = f/g * 10**(x.a -
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#     hi = +n * +recip # unary `+` chops to ctx.prec digits
+#
+# we have 3 visible chopped operationa, but there's also a 4th:
+# precomuting a truncatrd `recip` as part of setup.
```
```suggestion
# precomuting a truncated `recip` as part of setup.
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+# On Computing Reciprocals
+# ------------------------
+# In general, the exact reciprocals we compute have over twice as many
+# significant digts as needed. 1/256**i has the same number of
```
```suggestion
# significant digits as needed. 1/256**i has the same number of
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+# significant decimal digits as 5**i. It's a significant waste of RAM
+# to store all those unneeded digits.
+#
+# So we cut exact repicroals back to the least precision that can
```
```suggestion
# So we cut exact reciprocals back to the least precision that can
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+# p256.a + p256.a + 1 - 2 * p256.a = 1. QED
+#
+# Note: an earlier version of the code split on floor(e/2) instead of on
+# the ceiiing. Thw worst case then is odd `w`, and a more involved proof
```
```suggestion
# the ceiling. The worst case then is odd `w`, and a more involved proof
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+# Note: an earlier version of the code split on floor(e/2) instead of on
+# the ceiiing. Thw worst case then is odd `w`, and a more involved proof
+# was needed to show that adding 4 (instead of 1) may be neceasary.
+# Basically because, in that cass, n may be up to 256 times larger than
```
```suggestion
# Basically because, in that case, n may be up to 256 times larger than
```

nineteendo reviewed (commented):
I found some more typo's

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Apply suggestions from code review

Co-authored-by: Nice Zombies <nineteendo19d0@gmail.com>

tim-one committed:
Merge branch 'str2int' of https://github.com/tim-one/cpython into str2int

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+# GUARD additional digits.
+#
+# The default 8 is more than enough so no more than 1 correction step
+# was ever needed for all inputs tried through 2.5 billion digita. In
```
```suggestion
# was ever needed for all inputs tried through 2.5 billion digits. In
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#     hi = +n * +recip # unary `+` chops to ctx.prec digits
+#
+# we have 3 visible chopped operationa, but there's also a 4th:
+# precomuting a truncated `recip` as part of setup.
```
```suggestion
# precomputing a truncated `recip` as part of setup.
```

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#     n.a - 2 * p256.a <= 1
+#
+# What's the largest n can be? n < 255**w = 256**(w2 + (w - w2)). The
+# worst case in this context is when w ix even. and then w = 2*w2, ao
```
```suggestion
# worst case in this context is when w ix even. and then w = 2*w2, as
```

tim-one replied:
It's a typo, but "so" was intended, which I'll change.

nineteendo commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#
+# Note: an earlier version of the code split on floor(e/2) instead of on
+# the ceiling. The worst case then is odd `w`, and a more involved proof
+# was needed to show that adding 4 (instead of 1) may be neceasary.
```
```suggestion
# was needed to show that adding 4 (instead of 1) may be necessary.
```

nineteendo reviewed (commented):
Hmm, found some more

tim-one commented on Lib/_pylong.py:
```diff
@@ -361,3 +533,191 @@ def int_divmod(a, b):
+#     n.a - 2 * p256.a <= 1
+#
+# What's the largest n can be? n < 255**w = 256**(w2 + (w - w2)). The
+# worst case in this context is when w ix even. and then w = 2*w2, ao
```
```suggestion
# worst case in this context is when w ix even. and then w = 2*w2, so
```

tim-one committed:
Apply suggestions from code review

Co-authored-by: Nice Zombies <nineteendo19d0@gmail.com>

tim-one committed:
Merge branch 'str2int' of https://github.com/tim-one/cpython into str2int

tim-one committed:
Add limit=0 to compute_powers() "by hand" testing.
No particularly good reason to - it just "looked funny" that
0 _wasn't_ tested. It's not a realistic value to use, though.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Reduce cutoff for calling the new implementation from 3.5M to 2M.

It could probably be reduced more as-is, and more still with more
fine-tuning. This was remarkably easy: it just required inner()
to let Python convert much larger strings to ints, so large that
the existing Karatsuba int(str) usually gets invoked.

tim-one committed:
Update NEWS.

tim-one commented on Misc/NEWS.d/next/Core and Builtins/2024-05-09-02-37-25.gh-issue-118750.7aLfT-.rst:
```diff
@@ -0,0 +1 @@
+If the C version of the ``decimal`` module is available, ``int(str)`` now uses it to supply an asymptotically much faster conversion. However, this only applies if the string contains over about 3 million digits.
```
```suggestion
If the C version of the ``decimal`` module is available, ``int(str)`` now uses it to supply an asymptotically much faster conversion. However, this only applies if the string contains over about 2 million digits.
```

tim-one committed:
Sheesh - put a wrong number in the new NEWS.

tim-one committed:
Merge remote-tracking branch 'upstream/main' into str2int

tim-one committed:
Spell out the additional new reason not to do int(Decimal) directly.
